### DISCUSSION TITLE
Speeding up SSH connection setup for IOS-XR

### DISCUSSION THREAD

**Discussion by mpdavy** [ktbyers/netmiko#2342] [Q&A] [answered]
In version 3.4, the CiscoXrBase session_preparation method (cisco/cisco_xr.py) has a couple of self._test_channel_read() calls added with a note of:

IOS-XR has an issue where it echoes the command even though it hasn't returned the prompt

I'm curious which version(s) of XR this issue was seen with and how best to test if a router is exhibiting this behavior ?

These 2 calls to _test_channel_read() are adding about 5 seconds total to the SSH session setup for XR. I'm wondering if this can be sped up a bit assuming the specific XR version and platform we're using does not exhibit this behavior.

**Comment by ktbyers** [comment]
Okay, I think I broke something on the performance of IOS-XR between Netmiko version 3.3.2 and the current develop branch (likely also broke on Netmiko version 3.4.0).

I say this looking at my historical performance data for IOS-XR and comparing it to what I see for the current response time.

I will see if I can create a proposed fix for it and post the proposed fix here.

**Reply by ktbyers** [reply]
https://github.com/ktbyers/netmiko/issues/2343

**Comment by ktbyers** [comment]
Reference data for performance as of Netmiko 3.3.2:

https://raw.githubusercontent.com/ktbyers/netmiko/7b8b16517cba757bbb5925b43772d486775fb69b/tests/performance/netmiko_cisco_xr.svg

**Reply by ktbyers** [reply]
This should fix the issue:

https://github.com/ktbyers/netmiko/pull/2345

One problem, however, is that this will be with Netmiko 4.x release which won't come out for a while (some time in the second half of the year). The other temporary option would to downgrade to Netmiko 3.3.2.

Note, Netmiko 4.x also has a lot of other changes so you probably shouldn't directly use it now (without quite a bit of additional testing of your uses).

Does that work for you?

**Reply by mpdavy** [reply]
Thanks for looking into this.  I'll have to check.  We recently upgraded to 3.4 for some other fix/improvement.

**Reply by mpdavy** [reply]
Just wanted to follow-up and confirm that 3.3.2 resolved the issue.  Observed SSH connection setup times dropped from > 7 seconds in 3.4.0 to < 2 seconds in 3.3.2

**Comment by ktbyers** [comment, marked as answer]
Bug fixed in #2345
